### ISSUE TITLE
fbsimctl direct launch mode -- keeping process up

### ISSUE THREAD

**Issue by RavisMsk** [facebook/FBSimulatorControl#286]
Hello,

Got a question about direct launch mode. Usage documentation for fbsimctl states:
> However, fbsimctl will connect a Framebuffer to the Simulator as it is Booting. This enables Video Recording, providing that the fbsimctl process is kept alive.

What is the recommended way to keep fbsimctl process up? At the moment I'm trying to use `listen` for that purpose and launching it as a subprocess or in another thread. But the problem is that it says:
```
{"event_type":"discrete","subject":"Error Domain=com.facebook.FBSimulatorControl Code=0 \"Simulator Does not have a FBFramebufferVideo instance\" UserInfo={NSLocalizedDescription=Simulator Does not have a FBFramebufferVideo instance}","timestamp":1469720386,"event_name":"failure"}
```
for `record start`. Seems like it can't allocate Framebuffer on another thread/process.

There's also another question -- can multiple fbsimctl instances run at the same time (in case I'm trying to keep direct-launch mode up)? I'm having issues running 8+ simulators in parallel, it throws posix_spawn error 35, seems to be connected with some shared resources, but I'm not sure if its on the side of fbsimctl or CoreSimulator.

Thanks for great library and utilities, I'm trying to use it in parallel UI-tests automation scripts. Most valuable features for me are:
- direct-launch (seems to reduce resources requirements per simulator);
- video recording (it's way easier to find problems if test fails);
- watchdog override (on simulator.app my app gets killed sometimes during heavy loads);

**Comment by RavisMsk** [comment]
Forgot to mention that I've tried booting direct-launch mode with listen and using its stdio/stdout to send commands and receive events using `select` etc. That way it has allocated framebuffer, but that method is way more difficult and I'm having some issues with broken pipes etc.

**Comment by lawrencelomax** [comment]
> What is the recommended way to keep fbsimctl process up? At the moment I'm trying to use listen for that purpose and launching it as a subprocess or in another thread. But the problem is that it says

Yes, chaining with `listen` is currently the best way of doing this. You don't need to start a http server, `listen` with no args will just effectively keep the process up until you send a `SIG[TERM|HUP]`.

One day I'd like for a better way to keep a Framebuffer alive and speak to this process with some IPC to allow additional `fbsimctl` processes to manipulate it, but that's quite a lot of work. This means that currently of `record start` has to come from the *same* process that has booted with `--direct-launch`. This does mean that you can start recording from an external process by sending the recording start action over HTTP.

> There's also another question -- can multiple fbsimctl instances run at the same time (in case I'm trying to keep direct-launch mode up)? I'm having issues running 8+ simulators in parallel, it throws posix_spawn error 35, seems to be connected with some shared resources, but I'm not sure if its on the side of fbsimctl or CoreSimulator.

In my experience you'd start to see issues at around 8 simulators, 7 is safe though.

I've not dug in too much to why this number is the 'sweet spot', but I imagine this has a lot to do with the number of processes that are required for a Simulator to operate. The `launchd_sim` of a Simulator (the root process of a single simulator) will launch a bunch of daemon processes for iOS to run like `SpringBoard`, `backboardd`, `locationd` etc. etc.

You should look at reducing the scale with the `--scale=25|50|75` argument. This will pass in a *lower resolution* framebuffer on launch, which can reduce memory usage. There's less need to have a full resolution video, especially with the insane resolutions that iOS devices have these days. It will also mean less work for the video encoder.

> direct-launch (seems to reduce resources requirements per simulator)

I've also seen the same. It's likely a lot of this has to do with rendering the Simulator's Framebuffer inside a Mac OS X Application. Opening 8+ Applications will certainly put more strain on the system with an `NSApplication`/`NSWindow` and all the associated surfaces that are required to render the Simulator's frame.

**Comment by lawrencelomax** [comment]
> Forgot to mention that I've tried booting direct-launch mode with listen and using its stdio/stdout to send commands and receive events using select etc. That way it has allocated framebuffer, but that method is way more difficult and I'm having some issues with broken pipes etc.

I imagine that there's something going on with the implementation of this in`fbsimctl` so there's probably some fixes to be made there.

I didn't mention it yet, but there are also some changes that I'm hoping to land soon that will demonstrate how to treat recording in the [`cli-tests`](https://github.com/facebook/FBSimulatorControl/blob/master/fbsimctl/cli-tests/tests.py), however they don't demonstrate this with piping commands to `stdin`.

**Comment by RavisMsk** [comment]
@lawrencelomax thank you for operative response, you clarified some issues I had.
Now I'm sticking with piping method, seems to allocate framebuffer successfully, but requires interesting synchronization on io (for reads and writes) to avoid broken pipes etc.

Found out another non-obvious moment -- if you start multiple `fbsimctl` with `listen`s and write to their `stdin`s, then commands should still contain query. At first I thought that each `listen` is connected only to its queried devices, eg `fbsimctl <sim_id> boot --direct-launch -- listen` targets only `<sim_id>`. Maybe that should be clarified in documentation (or is it a bug?).

According to that (if it's not a bug), am I right, that proposed method for running `listen` is to start single instance with `listen` and then communicate with it for all required simulators? Because I was trying to fork `listen` processes for each parallel simulator.

**Comment by lawrencelomax** [comment]
I've just landed the [e2e tests for the cli that demonstrate video recording](https://github.com/facebook/FBSimulatorControl/blob/master/fbsimctl/cli-tests/tests.py#L170), these appear [to be working great in Travis too.](https://travis-ci.org/facebook/FBSimulatorControl/jobs/148259443#L2021).

> Now I'm sticking with piping method, seems to allocate framebuffer successfully, but requires interesting synchronization on io (for reads and writes) to avoid broken pipes etc.

Yikes, this sounds like something that needs fixing in `fbsimctl`.

> Found out another non-obvious moment -- if you start multiple fbsimctl with listens and write to their stdins, then commands should still contain query.

This also doesn't sound like the intended behavior. I would expect that you should be able to do the following:
```bash
fbsimctl UDID listen
# now listening on stdin
boot 
# simulator should boot
shutdown 
# simulator should shutdown
ANOTHER-UDID boot
# a different simulator boots
``

i.e the query passed to the `listen` should act as the default for any action you run without any query specified. Specifying a query when listening should override whatever you passed when you started the `listen`. I'll look at clarifying/fixing this.

> According to that (if it's not a bug), am I right, that proposed method for running listen is to start single instance with listen and then communicate with it for all required simulators? Because I was trying to fork listen processes for each parallel simulator

I wouldn't use this method to co-ordinate and synchronize between multiple simulators. I'd look at using multiple device sets, specified by the `--set` argument. This will allow you to specify a path to a custom Simulator set.

With this you can have concurrent supervisor processes of `fbsimctl` each with their own `listen` command, each managing a single simulator. This reduces complexity as you won't have to synchronize multiple simulators across one listening `fbsimctl` process. You also have much better isolation between the supervisor processes. It also means you can manage at the set level, rather than at the Simulator level.

I hope that helps!

**Comment by lawrencelomax** [comment]
I'm going to close this for now, if you could re-open another issue with a report for the specific `listen` issue that you have that would be awesome 👍
